Add support to use arrays as names of objects

sodevel opened the pull request:
This allows to use names of the form myWidget[0], myWidget[1], ... for objects, multiple dimensions are also allowed. Additionally for C++ an anonymous enum with symbols of the form MYWIDGET_SIZE for single dimensional arrays and MYWIDGET_0_SIZE, MYWIDGET_1_SIZE, ... for multi dimensional arrays are created which store the number of elements in the corresponding dimension.

This change is local to the C++ code generator with one catch that i will explain below.

This is a minimal-change-approach implementation that bases on the fact that declaring a static array and accessing its members is syntactically identical of using a regular member, only special care has to be taken during declaration of the array to declare it only once with the correct dimensions.

The implementation works like this:
- In a preprocessing step for each top-level object parse its object tree and extract the name of every found array and record the highest used index
- The two locations that declare an object compare the name of the current object against the list of array names
- If the name does not match an array continue as usual, otherwise
- If the array has already been declared skip the object, otherwise
- Create the enum values with the count of the elements in the dimensions
- Temporary rename the object to contain the count of elements instead the current index
- Call the declaration template
- Restore the name of the object, mark the array as declared and continue as usual

Now the catch, certain code templates, e.g. for a wxChoice, contain fragments like this:
`wxArrayString $name #append data;`
This will result in an invalid identifier if `$name` contains array brackets `[]`. The same approach like before, renaming the object to something else, doesn't work here because there are other locations in the template that actually need the name with the array index. To solve this, the `#append` function searches for the beginning of its preceeding token and replaces `[]` with `_`.

The template processor however is generic code, it is used by all targets so this will affect the other generators as well. I checked all templates and this change *should* not break anything, but i am not totally sure because i don't know enough of the other languages. And this is also the reason why this is only implemented in C++, i am unable to implement this for the other languages and don't know if this is even possible or necessary at all, maybe some smart tricks work there out-of-the-box.

This feature however is very important for me and one of the reasons why i have a not-so-private-anymore fork of wxFormBuilder, so if it won't make it into the official project again (it got rejected in the SourceForge days), maybe others can help to implement this for the other target languages to get a better chance to get it accepted.

jhasse commented on src/codegen/cppcg.cpp:
```diff
@@ -1096,7 +1099,76 @@ wxString CppCodeGenerator::GetCode( PObjectBase obj, wxString name )
+
+	// Get the name
+	auto propName = obj->GetProperty(wxT("name"));
+	if (propName)
```
Use an early return to avoid the indentation:

```cpp
if (!propName) {
		// Object has no name, just get its code
		return GetCode(obj, wxT("declaration"));
}
```

jhasse commented on src/codegen/cppcg.cpp:
```diff
@@ -1096,7 +1099,76 @@ wxString CppCodeGenerator::GetCode( PObjectBase obj, wxString name )
+		wxString baseName;
+		ArrayIndex index;
+
+		if (ParseArrayName(name, baseName, index))
```
Also here.

jhasse commented on src/codegen/cppcg.cpp:
```diff
@@ -1096,7 +1099,76 @@ wxString CppCodeGenerator::GetCode( PObjectBase obj, wxString name )
+		{
+			// Object is an array, check if it needs to be declared
+			auto& item = arrays[baseName];
+			if (!item.second)
```
And here.

jhasse commented:
What are the enums for?

sodevel commented on src/codegen/cppcg.h:
```diff
@@ -73,6 +75,21 @@ class CppCodeGenerator : public CodeGenerator
+	* Map of all discovered array items.
+	* The key element is the basename of the array, the value the describing ArrayItem.
+	*/
+	typedef std::map<wxString, ArrayItem> ArrayMap;
```
The code evolved over time, i will try to find some nicer names

sodevel commented on src/codegen/cppcg.cpp:
```diff
@@ -1824,6 +1896,30 @@ void CppCodeGenerator::FindEventHandlers( PObjectBase obj, EventVector &events )
+	for (
+		auto indexStart = name.find_first_of(wxT('[')), indexEnd = name.find_first_of(wxT(']'), indexStart + 1);
+		indexStart != wxString::npos && indexEnd != wxString::npos;
+		indexStart = name.find_first_of(wxT('['), indexEnd + 1), indexEnd = name.find_first_of(wxT(']'), indexStart + 1)
```
Lines cost money, need to write short code :smile:. I will take a look at this.

jhasse replied:
My version has more lines, but the lines are shorter ;)

sodevel commented:
To be able to easily enumerate over the array using an index. I have lots of old code that doesn't use ranged-based for loops and often i do need the index to enumerate two containers in sync.

jhasse commented:
What about using `sizeof(arr) / sizeof(arr[0])` or `std::array` instead?

sodevel commented:
I really like to store constant values like `sizeof(arr) / sizeof(arr[0])` in an enum, just to give them a meaningful name and without the need to change formulas like this all over the place if they needs to be updated. Maybe i could add a project property to disable generation of the enum values.

Now the idea with `std::array` sounds nice, but this would make the generated code require c++11 and i don't think this is currently required, not sure if this requirement should be added just for this. But the bigger issue is, this doesn't work with the "trick" the array support is based on. The declaration code templates look like this:
`#class* $name;`
To create a static array, it's enough to make `$name` look like one, but you can't create a `std::array` like this. The good thing is that the `#class` macro calls into the code generator class for the target language, the bad thing is the `*` isn't part of it. So even if `#class` knows how to create a `std::array` it would still produce an invalid result. Maybe the `#class` macro can be changed like the `#append` macro to include some adjacent characters, but this again would be in the generic code which affects all target languages. I don't know enough about the syntax of the other languages, and c++ allows even more complicate syntax like whitespace in front of the `*`, this requires more research to not break things.

sodevel commented:
Rebased on current master incorporating the suggested code updates. Added a project property to control the generation of the array enum values and always skip them for private objects.

sodevel commented:
I'm trying to reduce my rebases and forced pushes until this reaches the finish line, can do a final cleanup when the time has come.

jhasse commented:
I see! Thanks for the explanation. As I understand it (since the declaration is a pointer) my sizeof trick doesn't work anyway.

sodevel commented:
I have added implementations for the remaining languages.

Disclaimer: I don't know these languages, i've read some documenation and the generated code looks correct for me, but i don't know if it really is. Would be nice if someone could verify that.

The Python implementation is the only one that does only support one dimensional arrays, i had to use dictionaries to be able to easily add elements using the array notation. Apparently it is not possible to put dictionaries into dictionaries so i don't know how to realize multidimensional arrays like that. I don't know if anyone ever needs multi dimensional arrays 😄

sodevel commented:
Ok, i guess clicking the "Resolve Conflicts" button was not a good idea, i didn't plan to pull in current master.

sodevel commented:
Playing around with online compilers i discovered the Lua implementation currently also doesn't support multi dimensional arrays.

Technically it is possible to support multi dimensional arrays for Python and Lua, the dictionaries / tables need to be populated manually with the required count of elements in all dimensions but the last.

sodevel commented:
I updated the code to properly create multi dimensional arrays for Python and Lua. I tested the generated array code of all languages with online compilers and it works correct. This feature is now complete and all languages have the same features with the small exception that C++ has these dimension enum's. The other languages don't need this because they use smarter "arrays" like the plain C++ arrays which have integrated enumeration features (and PHP and Lua don't have enums at all).

I forced push a huge cleanup with a better history than the previous commits.

jhasse commented:
Sorry for the late update. I'll have a more detailed look at this soon.

sodevel commented on src/codegen/codegen.cpp:
```diff
@@ -406,14 +409,16 @@ bool TemplateParser::ParseForm()
-	PProperty propNs= project->GetProperty( wxT( "ui_table" ) );
-	if ( propNs )
+	const auto& project = AppData()->GetProjectData();
+	const auto& table = project->GetProperty(wxT("ui_table"));
```
No, the const& will extend its lifetime.

I'm still struggling whats the best way to store method results in a performant, refactoring friendly, uniform manner. If i don't want to modify the result i tend to use const auto& unless i know for sure the returned element is a primitive, then i use const auto. For iterators i use plain auto.

With const auto& you can bind anything and prevent creating copies if the method does return a reference, it is unnecessary though if it does return a temporary like here, these methods both return a shared_ptr.

jhasse replied:
> No, the const& will extend its lifetime.

Ah right! C++ is full of surprises ;)

Functions shouldn't return references, therefore I would always use `const auto`.

sodevel commented on src/codegen/cppcg.cpp:
```diff
@@ -1576,7 +1662,7 @@ void CppCodeGenerator::GenConstruction( PObjectBase obj, bool is_widget )
 		if ( perm_str == wxT( "none" ) )
 		{
-			const wxString& decl = GetCode( obj, wxT( "declaration" ) );
+			const auto& decl = GetDeclaration(obj, arrays, false);
```
I don't see a difference to the other locations where the documentation gets moved into the auto

jhasse replied:
Yes, I think it would also be good at the other locations. But you're right that `auto` makes it more refactor-friendly, so never mind.

sodevel commented:
Half a year has passed, can i assume this wont make it into master? Im just asking because now i do need a new wxFormBuilder build because of #499 so i need to find a solution to manage my own, private builds again.

jhasse commented:
Sorry, I forgot about this. You seem to be using this already, so I'll will merge it.

You're doing a lot of work on wxFormBuilder, so I think it makes sense that you gain commit rights to this repo. Would you be okay?

sodevel commented:
Thank you for the offer however i can work on this project only as my schedule allows. We are using this application at the company i work for and i mostly work on it if i need to fix or add something that we do need. So i can't promise regular contributions from my side.
